Execute Animated.js declarative animation on UIThread on Android.

kmagiera opened the pull request:
This is the first from the series of PRs I'm going to be sending shorty that would let Animated.js animations to run off the JS thread (for Android only).


This PR introduce a new native module that will be used for offloading animations - NativeAnimatedModule. It has a simple API that allows for animated nodes management via methods like: create/drop animated node, connect/disconnect nodes, start animation of a value node, attach/detach animated from a native view.


Similarly to how we handle UIManager view hierarchy updates we create a queue of animated graph operations that are then executed on the UI thread. This isolates us from problems that may be caused by concurrent updates of animated graph while UI thread is "executing" the animation.


The most important class NativeAnimatedNodesManager.java implements a management interface for animated nodes graph as well as implements a graph traversal algorithm that is run for each animation frame. For each animation frame we visit animated nodes that might've been updated as well as their children that may use parent's values to update themselves. At the end of the traversal algorithm we expect to reach a special type of the node: PropsAnimatedNode that is then responsible for calculating property map which can be sent to native view hierarchy to update the view.


From the "client" perspective the API is as simple as adding: `useNativeDriver: true` to the set of animation properties, e.g:
```
Animated.timing(
  this.state.fadeAnim,
  {
    toValue: 1,
    duration: 2000,
    useNativeDriver: true,
  },
).start();
```


This approach doesn't let us support animating layout properties as it updates the values directly on the native view hierarchy. In order to be able to handle updates of layout props we'd need to run the updates on shadow view hierarchy first (on "native" thread). However a lot of this code could be reused if we decided to implement support for layout props (my initial prototype was capable of updating layout props off the JS thread).


This PR adds support only for a limited number of Animated.js library features (mainly to keep it shorter), here is what's supported as of this PR:
 - Works only for android
 - Only one type of animation: Animated.timing
 - One can use animated values only for non-layout, numeric properties (e.g. support for transform property is disabled)
 - One can use JS driven version of Animated.js library along with "native" animated. The only limitation is that we cannot drive same value both by native and JS animated driver.


What's coming up next (have most of the work done but will be sending this later to keep this change as simple as possible):
 - More animation types: decay, spring
 - Support for transform property
 - Support for "operation" nodes: Animated.add, Animated.multiply, etc.
 - Animated.event support
 - Unit tests for the graph traversal algorithm


**IMPORTANT**: This PR doesn't enable animated native module by default. In order to use it you'd need to add `NativeAnimatedModule` to the list of yours app 'packages'


**Test plan**
 1. Add `NativeAnimatedModule` to the list of native modules for you app
 2. Add `useNativeDriver: true` to your animation config (Animated.timing only for now). You can use my example app: https://gist.github.com/kmagiera/bbfbac6c00077d188bc5
 3. Enable SPY_MODE in MessageQueue.js to see that setNativeProps operation are no longer passed over the bridge while animation is running and `useNativeDriver` is set to `true`.

kmagiera commented:
cc: @astreet

mkonicek commented:
Wow thanks Krzysztof for sending this! :) :+1: :+1: :+1:

sahrens commented:
It's happening!

![](http://www.reactiongifs.com/r/swih.gif)

kmagiera commented on Libraries/Animated/src/AnimatedImplementation.js:
```diff
@@ -169,12 +279,28 @@ class TimingAnimation extends Animation {
     this._delay = config.delay !== undefined ? config.delay : 0;
     this.__isInteraction = config.isInteraction !== undefined ? config.isInteraction : true;
+    this._useNativeDriver = !!config.useNativeDriver;
+  }
```
wat?

vjeux replied:
I have no idea what's going on here, sorry for the spam :(

kmagiera replied:
looks like comments has been left on incorrect lines by the bot. I ended up running eslint locally to get the correct line numbers and fixed the issues cc @mkonicek

kmagiera commented:
I've fixed all the lint and flow error in my last update

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java:
```diff
@@ -83,6 +83,14 @@ public UIManagerModule(
+   * operations executed on that object should be run in the UI thread. Use with caution
+   */
+  public NativeViewHierarchyManager getNativeViewHierarchyManager() {
+    return mUIImplementation.getUIViewOperationQueue().getNativeViewHierarchyManager();
```
Not really a fan of exposing these objects. I think it might make more sense to integrate this all directly into UIManagerModule: there is definitely already implicit coupling (e.g. the ordering of callbacks in ReactChoreographer)

kmagiera replied:
Not happy with that approach either. I used to have a different way of setting up `NativeAnimatedModule` but decided to go with this instead. Here are some of the aspects driving the decision:
 1. I like the idea of using native module abstraction to expose API for managing animated nodes, starting/stopping animation etc. In my first prototype I put all those methods to UIManager. As it is already overloaded with a number of different methods (even `scrollTo` used to be a method of `UIManager`) I believe it's better for the animated native API to be extracted away from the UIManager
 2. Since animated API is extracted as a separate module (see (1)) I though it would be a good idea to make it "optional" so that developer can decide whether to include it on the RN "package" list (note that this module will be bringing "rebound" library as a dependency)

My "old" approach was to instantiate NativeAnimatedModule inside of UIManager and pass NativeHierarchyManager as a constructor argument to it. See corresponding changes in  [`UIManagerModule.java`](https://github.com/kmagiera/react-native/blob/1e9333d668a1959b180403cf3cd6267c4f01e67a/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java#L89). The issue with this approach was that it would require for the animated module to be added to the `CoreModulesPackage` (see it [here](https://github.com/kmagiera/react-native/blob/1e9333d668a1959b180403cf3cd6267c4f01e67a/ReactAndroid/src/main/java/com/facebook/react/CoreModulesPackage.java#L87)) and therefore would prevent us from making that module "optional" (If you think it should be a "core" module I'm fine with that too, just didn't want to make the decision for you).

Last thing re alternative implementations of `UIImplementation` - I didn't forget about it, just didn't know its still a thing. However, in order to drive animations from UI thread we need to have a synchronous way of updating the UI while `UIImplementation` is asynchronous (at least the default implementation of it). Don't have any good ideas on how can this be solved. One thing that comes to my mind is to add method to `UIImplementation` - `synchronouslyUpdateViewOnUIThread` as that's the only method we'll ever need to access from the perspective of native animated module.

astreet replied:
Ok, I feel convinced that this should be a separate module (if only because of Rebound). Exposing the NativeViewHierarchyManager seems like the only reasonable way to do this that I can think of unfortunately...

astreet replied:
I also let Denis Koroskin + Nodes know about this PR, so hopefully one of them comments

vjeux replied:
We should probably not include rebound but instead just inline the spring implementation like I did: https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/AnimatedImplementation.js#L302-L514

Rebound comes with a bunch of utilities and classes that we don't need for Animated. The core "physics" is actually just a loop with only a dozen of lines of code.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/AnimatedNode.java:
```diff
@@ -0,0 +1,81 @@
+  private static final int DEFAULT_ANIMATED_NODE_CHILD_COUNT = 1;
+  /*package*/ static final int INITIAL_DFS_COLOR = 0;
+
+  /*package*/ @Nullable List<AnimatedNode> mChildren; /* lazy-initialized when a child is added */
```
... we should probably be doing this in CSSNode

astreet commented:
The general idea of running animations on the main thread by interacting with the NativeViewHierarchyManager sounds like the right approach here. Couple things:

1. We have an alternate implementation of UIImplementation internally (called 'nodes' if you remember it) -- it extends UIImplementation and overrides things like handleUpdateView and handleCreateView. We also have another implementation of NativeViewHierarchyManager (that extends NativeViewHierarchyManager, adding some methods that the other UIImplementation uses). Just giving you context because hopefully we make something that just works for both (e.g., it shouldn't matter what the UI is implemented with since both implementations respect referencing tags and setting props)

2. Can you explain why running animations on the main thread requires this node graph/resolution algorithm? Maybe an example would be helpful.

Also happy to talk on Skype or something instead of communicating over text :)

kmagiera commented:
Re 1) I responded inline


Re 2) this is just a mapping of what AnimatedImplementation.js is doing to drive animations in JS. I'll try to explain it in few words:


The animated js is based on a concept of a graph where nodes are values or transform operations (such as interpolation, addition, etc) and connection are used to describe how change of the value in one node can affect other nodes. Few examples:
 - `Animated.Value` is a simplest type of node with numeric value which can be driven by an animation engine (spring, decay, etc) or by calling `setValue` on it from JS
 - `Animated.add` is a type of node that may have two or more input nodes. It outputs the sum of all the input node values
 - `interpolate` - is actually a method you can call on any node and it creates a new node that takes the parent node as an input and outputs its interpolated value (e.g. if you have value that can animate from 0 to 1 you can create interpolated node and set output range to be 0 to 100 and when the input node changes the output of interpolated node will be multiplying the values by 100)


You can mix and chain nodes however you like and this way create nodes graph with connections between them.


To map animated node values to view properties there is a special type of a node: [`AnimatedProps`](https://github.com/facebook/react-native/blob/593d766ec57ec2b770dde68320008f56776cd6ce/Libraries/Animated/src/AnimatedImplementation.js#L1049). It is created by `AnimatedImplementation` whenever you render `Animated.View` and stores a mapping from the view properties to the corresponding animated values (so it's actually also a node with connections to the value nodes). Example:

```
var someAnimatedValue = new Animated.Value(0);

...

<Animated.View
  style={{opacity: someAnimatedValue}}>
  <Text>Stuff</Text>
</Animated.View>
```

The code from the above will generate 3 nodes:
 - [`AnimatedValue`](https://github.com/facebook/react-native/blob/593d766ec57ec2b770dde68320008f56776cd6ce/Libraries/Animated/src/AnimatedImplementation.js#L526) node with a value set to 0
 - [`AnimatedStyle`](https://github.com/facebook/react-native/blob/593d766ec57ec2b770dde68320008f56776cd6ce/Libraries/Animated/src/AnimatedImplementation.js#L991) node that stores a mapping from property with the name "opacity" to the value node mentioned above
 - [`AnimatedProps`](https://github.com/facebook/react-native/blob/593d766ec57ec2b770dde68320008f56776cd6ce/Libraries/Animated/src/AnimatedImplementation.js#L1049) node with style property being mapped to style node from the above


Last "special" parts of the the graph are "animation drivers". Those are objects that based on some criteria updates attached values every frame (we have few types of those, e.g., spring, timing, decay). Animation objects can be "started" and "stopped". In JS they use `setTimeout` to run value update loop every frame. They are like "pulse generators" for the rest of the nodes graph. Those pulses then propagate along the graph from nodes to their children up to the `AnimatedProps` nodes. Example:

```
Animated.timing(
  thatAnimatedValueFromThePreviousSnippet,
  {
    toValue: 1,
    duration: 500,
  },
 ).start(); 
```


The code above will create an object that will be responsible for updating `thatAnimatedValueFromThePreviousSnippet` animated value from their current state at the moment `start` is called (that would be 0) to the value specified in the animation config (`toValue: 1`) in a linear manner every frame for 500 milliseconds (`duration: 500`)


The process of calculating view updates during animation works as follows (I'm describing the process from the perspective of Animated JS and how it runs in javascript, but it works very similar in native):
 1. You run all the active animations that may update attached animated value nodes
 2. For each updated node you visit their children nodes as they may also require an update
 3. Once you reach `AnimatedProps` node you can collect properties which needs to be updated, then we send `setNativeProps` to the corresponding view


So in order to calculate property updates you need to now how is the value generated by the animation driver (e.g. spring) processed before it gets transformed into a value that can be set as a property on a view. This is why I'm using the simplest possible approach and just send and maintain that animated nodes graph in native. There are definitely other ways of handling that, but I believe this one is the most flexible method as it naturally reflects the public animated.js API. It is also future proof as far as the API isn't going to change dramatically and only new type of nodes are going to be added.

astreet commented:
Ok, got it, thanks for the explanation. I'm guessing we're just keeping the same architecture in native? The one thing I didn't understand was 

>The code from the above will generate 3 nodes:

> - AnimatedValue node with a value set to 0
> - AnimatedStyle node that stores a mapping from property with the name "opacity" to the value node mentioned above
> - AnimatedProps node with style property being mapped to style node from the above

What is the purpose of having both AnimatedStyle and AnimatedProps nodes?

I think I mostly have enough context to review this now, I will try to work on it tonight or tomorrow (currently traveling in Tel Aviv so my hours are a bit weird)

kmagiera commented:
> I'm guessing we're just keeping the same architecture in native? The one thing I didn't understand was

Yes, As I mentioned earlier, IMO this will allow us to easier adapt to any changes in Animated.JS (like new node types etc).


> What is the purpose of having both AnimatedStyle and AnimatedProps nodes?

That's a very good question I don't really know the answer for. So I'm just mapping this to native as it supporting separate graph structure would make it all more complex. I'm guessing this has been designed the way just to split up handling nested objects into an existing node-with-children abstraction (we do similar thing for other non-scalar properties such as transform - there is [`AnimatedTransform`](https://github.com/facebook/react-native/blob/593d766ec57ec2b770dde68320008f56776cd6ce/Libraries/Animated/src/AnimatedImplementation.js#L929) node) . Note that for an `Animated.View` you can animate any property, not only style props (doesn't make sense most of the time though). Also, from the perspective of Animated.JS, we don't  care that all styles props are just flattened and sent over the bridge as they were direct properties of the view - so the structure of all the props needs to be maintained (the flattening "magic" happens [here](https://github.com/facebook/react-native/blob/ad8a33586410c6f9048983f64f8f86e0715e73b8/Libraries/ReactIOS/NativeMethodsMixin.js#L147)). In fact, I'd prefer for all those styles, props and transform to be mapped to a single graph node. As at the moment those are the only nodes that outputs non-scalar values (maps with stuff), so having them in a single node would simplify interface we use for the nodes to communicate. 

Perhaps @vjeux can answer this better or make a suggestion whether it makes sense to join Style, Props and AnimatedTransform together.

vjeux commented:
> What is the purpose of having both AnimatedStyle and AnimatedProps nodes?

Mix of legacy and easier to manage. I started with AnimatedStyle to represent the style object, but then realized that I needed to do something very similar for transform, so I copy and pasted it to AnimatedTransform. Then, someone wanted to animate other attributes than styles (for ART), so I copy and pasted it another time with AnimatedProps.

I have no issue if you want to flatten the three into a single AnimatedNode.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/AnimatedNode.java:
```diff
@@ -0,0 +1,81 @@
+  /*package*/ int mDFSColor = INITIAL_DFS_COLOR;
+  public int mTag = -1;
+
+  double mValue = Double.NaN;
```
nit: group package and public vars together, add package qualifier to mValue

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/AnimatedNode.java:
```diff
@@ -0,0 +1,81 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.react.uimanager.animation;
```
I'd prefer if we create a new package for this module/PR

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/AnimatedNode.java:
```diff
@@ -0,0 +1,81 @@
+import javax.annotation.Nullable;
+
+/**
+ * Base class for all Animated.js library node types that can be created on the "native" side.
```
It'd be nice to add an explanation of how this is basically to represent a dependency DAG where child nodes depend on their parent nodes values to calculate their own value

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/AnimatedNode.java:
```diff
@@ -0,0 +1,81 @@
+
+  /**
+   * This method will be run in animation loop once all the node's predecessor nodes have been
+   * visited. This method can be overridden by subclasses to support use-cases when taking the input
```
Shouldn't we just only call this once all parents have been calculated?

kmagiera replied:
Yes, that's what it is used for

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/FrameBasedAnimationDriver.java:
```diff
@@ -0,0 +1,76 @@
+    long timeFromStartNanos = (frameTimeNanos - mStartFrameTimeNanos);
+    // frames are calculated at 60FPS, to get index by a given time offset from the start of the
+    // animation, we take the time diff in millisecond and divide it by 60 frames per 1000ms.
+    int frameIndex = (int) (timeFromStartNanos / 1000000L * 60L / 1000L);
```
This probably doesn't matter for this diff since we probably won't have animations that run long enough, but this isn't necessarily safe. Phones run at **about** 60FPS, but can vary (e.g. I think my Nexus 5 runs at ~59FPS optimally). For our internal framerate logging, we generally keep track of when we saw the last frame and use that to try to calculate how many frames occurred since then by dividing by the estimated frame time (1000 / 60ms). This will generally keep you from accumulating errors over time that you get by assuming exactly 60FPS.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/FrameBasedAnimationDriver.java:
```diff
@@ -0,0 +1,76 @@
+    int frameIndex = (int) (timeFromStartNanos / 1000000L * 60L / 1000L);
+    if (frameIndex < 0) {
+      throw new IllegalStateException("Calculated frame index should never be lower than 0");
+    } else if (!mHasFinished) {
```
nit: don't need else

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedModule.java:
```diff
@@ -0,0 +1,228 @@
+  private ArrayList<UIThreadOperation> mOperations = new ArrayList<>();
+  private volatile @Nullable ArrayList<UIThreadOperation> mReadyOperations = null;
+
+  /*
```
nit: // comment

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactChoreographer.java:
```diff
@@ -38,9 +38,14 @@
     /**
+     * For use by {@link com.facebook.react.uimanager.animation.NativeAnimatedModule}
+     */
+    ANIMATIONS(2),
```
If this is all this is for, then call it NATIVE_ANIMATED_MODULE

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedModule.java:
```diff
@@ -0,0 +1,228 @@
+
+  @Override
+  public void onHostDestroy() {
+    clearFrameCallback();
```
is onHostPause sometimes not called?

kmagiera replied:
I hope it gets always called before onHostDestroy :)
`clearFrameCallback` is idempotent so I guess it doesn't matter

astreet replied:
I think we should just remove it from onHostDestroy then -- it's confusing otherwise and we shouldn't really be trying to handle cases that we don't think are correct to happen in the first place

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+import java.util.Queue;
+
+/**
+ * Should be accessed only from the UI Thread
```
More javadocs about what this class does

astreet replied:
It'd be nice if you provided an explanation like you did on the PR about what this graph of nodes represents

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+    }
+    PropsAnimatedNode propsAnimatedNode = (PropsAnimatedNode) node;
+    if (propsAnimatedNode.mConnectedViewTag != -1) {
+      throw new JSApplicationIllegalArgumentException("ANimated node " + animatedNodeTag + " is " +
```
nit: Animated

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+        "of type " + PropsAnimatedNode.class.getName());
+    }
+    PropsAnimatedNode propsAnimatedNode = (PropsAnimatedNode) node;
+    if (propsAnimatedNode.mConnectedViewTag == viewTag) {
```
Should probably throw if this isn't the case to catch bugs

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+   * {@code mAnimatedGraphDFSColor} to mark nodes as visited in each of the DFSes which saves
+   * additional loops for clearing "visited" states.
+   *
+   * First DFS starts with nodes that are in {@code mUpdatedNodes} or directly attached to an active
```
Comment what it means for a node to be in mUpdatedNodes. It's not obvious to me so I'm going to have to go read the code to find out

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+
+    // STEP 1.
+    // DFS over graph of nodes starting from ones from `mUpdatedNodes` and ones that are attached to
+    // active animations (from `mActiveANimations`. Update `mIncomingNodes` attribute for each node
```
nit: Animations

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+    for (int i = 0; i < mActiveAnimations.size(); i++) {
+      AnimationDriver animation = mActiveAnimations.get(i);
+      animation.runAnimationStep(frameTimeNanos);
+      AnimatedNode valueNode = animation.mAnimatedValue;
```
Not sure how possible this is, but what is expected if JS had called setAnimatedValue on a node that was also attached to an animation? In this case, we'd just overwrite that value.

kmagiera replied:
IMO there is no use-case for such an action. Currently it will not be possible as `setValue` from JS will actually stop active animation (if any) (see [AnimatedImplementation.js:553](https://github.com/facebook/react-native/blob/11985d5c7721b0f9c9f46a04515799db0eeb44eb/Libraries/Animated/src/AnimatedImplementation.js#L553))

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+
+    for (int i = 0; i < mActiveAnimations.size(); i++) {
+      AnimationDriver animation = mActiveAnimations.get(i);
+      animation.runAnimationStep(frameTimeNanos);
```
looks like we ignore the return value whenever this is called?

kmagiera replied:
goo catch. The return value was used to indicate if the value has changed. After refactoring I'm no longer using that information (I though it could've been used for optimising traversal but I believe it's not really worth doing as most of the animated graphs aren't very huge)

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+      AnimationDriver animation = mActiveAnimations.get(i);
+      animation.runAnimationStep(frameTimeNanos);
+      AnimatedNode valueNode = animation.mAnimatedValue;
+      if (valueNode.mDFSColor != mAnimatedGraphDFSColor) {
```
I assume this would only happen in the case I mentioned above (when setAnimatedValue was also called)

kmagiera replied:
It can also happen when the view is created for the first time (as it usually have a value assigned)

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    // STEP 2
+    // DFS over the graph of active nodes in topological order -> visit node only when its all
```
nit: all its

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+    // `mActiveIncomingNodes = 0` (those can only be the ones that we start DFS in the previous
+    // step). We store number of visited nodes in this step in `updatedNodesCount`
+
+    nodesQueue.clear();
```
unnecessary I'd hope

kmagiera replied:
yeah, by the time we reach this line queue should be empty

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+    ArrayList<PropsAnimatedNode> updatedPropNodes = new ArrayList<>();
+    while (!nodesQueue.isEmpty()) {
+      AnimatedNode nextNode = nodesQueue.poll();
+      nextNode.runAnimationStep(frameTimeNanos);
```
Seems like we should only allow drivers to drive animation values: can we just call this 'update', not pass the frame time, and make the contract that the node just uses its incoming nodes to calculate its new value?

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+      if (nextNode.mChildren != null) {
+        for (int i = 0; i < nextNode.mChildren.size(); i++) {
+          AnimatedNode child = nextNode.mChildren.get(i);
+          child.feedDataFromUpdatedParent(nextNode);
```
Why do nodes need this? Seems simpler to just have the runAnimationStep/update call from above that's only called once all parents have updated. I guess that'd require keeping links parent <--> child instead of just parent --> child, but feedDataFromUpdatedParent is kinda weird

vjeux replied:
I've done many different implementations of this traversal and the best one I found was the following:

- We update all the root nodes (Animated.Value) that are currently animated to their new values. eg: 0.2 -> 0.4.
- We traverse all the children that depends on those values recursively and mark all the leaf nodes (AnimatedProps) as needed to be updated.
- When we collected all the leaf nodes, we call update() on them. It asks their parent(s) what their current value are and compute their new value.
- Except for the root nodes (Animated.Value), none of the nodes are stateful. All the values are re-computed recursively from the root on every frame.

See this comment for a hopefully good explanation of the process and the code supporting it:

https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/AnimatedImplementation.js#L101-L135

vjeux replied:
The main reason doing a top down implementation doesn't work is that you can have values that depend on multiple parents. For example `transform: [{translateX: x}, {translateY: y}]`. In order to compute the AnimatedTransform value, you need to know x and y. But x and y may both be updated during a single frame. So you can run into issues where you update order is [x, transform, y] and transform has a wrong value.

The code here solves this by using a topological sort in order to avoid having this ordering issue. I think that it should be safe as well.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+          AnimatedNode child = nextNode.mChildren.get(i);
+          child.feedDataFromUpdatedParent(nextNode);
+          child.mActiveIncomingNodes--;
+          if (child.mDFSColor != mAnimatedGraphDFSColor && child.mActiveIncomingNodes == 0) {
```
We shouldn't have to check color here, right?

kmagiera replied:
I use colors here to avoid enqueueing same node more than once (it could happen in case when setValue is called more than once for a single value in one JS batch). Another example would be if you have node that is driven by animation driver and has children that you call setValue on.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,344 @@
+      PropsAnimatedNode propNode = updatedPropNodes.get(i);
+      PropsAnimatedNode.UpdateViewData data = propNode.createUpdateViewData();
+      if (data.mViewTag > 0) {
+        mEnqueuedUpdates.add(propNode.createUpdateViewData());
```
Why not just call nativeViewHierarchyManager.updateProperties here? Instead of creating a list and only using it in runUpdates?

kmagiera replied:
That should be fine. I'll update that. I used to run this code on "native thread". That's why I needed a queue here, but it's no longer necessary

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/StyleAnimatedNode.java:
```diff
@@ -0,0 +1,52 @@
+
+  @Override
+  public void saveInPropMap(String key, JavaOnlyMap propsMap) {
+  /* ignore key, style names are flattened */
```
nit: //

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/StyleAnimatedNode.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  @Override
+  public void saveInPropMap(String key, JavaOnlyMap propsMap) {
```
this just seems like a copy of PropsAnimatedNode

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/AnimatedNode.java:
```diff
@@ -0,0 +1,81 @@
+   * @param propsMap view property map that will be used in the view update. This method should
+   *                 store its value in this map
+   */
+  public void saveInPropMap(String key, JavaOnlyMap propsMap) {
```
I'd rather this not be a method on AnimatedNode. Seems like a parent can just traverse its children and store them in the map. AnimatedNode shouldn't need to know about storing things in a map (Only Prop/Style nodes need to know about that)

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/PropsAnimatedNode.java:
```diff
@@ -0,0 +1,66 @@
+ * the nodes of this type in {@link NativeAnimatedNodesManager} that is responsible for extracting
+ * a map of updated properties, which can be then passed down to the view.
+ */
+/*package*/ class PropsAnimatedNode extends AnimatedNode {
```
Seems kinda weird that this and Style node below have mValue members that wouldn't be used. Maybe we should create a separate parent class for those internal nodes (ValueAnimatedNode, AddAnimatedNode, etc.) that should actually be tracking a value

astreet commented:
Nice! Added comments inline, native approach looks good, kinda glanced at the JS but would prefer @vjeux or @sahrens to review.

astreet commented:
Also, how are we ensuring that devs are only trying to use the native driver for supported props? Seems like we should have a whitelist, or a good error message if we try to update a prop that doesn't exist on the native view.

vjeux commented on Libraries/Animated/src/AnimatedImplementation.js:
```diff
@@ -50,20 +122,36 @@ type AnimationConfig = {
+      this.__nativeTag,
+      animatedValue.__getNativeTag(),
+      this._getNativeAnimationConfig(),
+      this.__debouncedOnEnd.bind(this));
```
nit: `);` on the next line

vjeux commented on Libraries/Animated/src/AnimatedImplementation.js:
```diff
@@ -1100,17 +1277,78 @@ class AnimatedProps extends Animated {
+    var nativeViewTag: ?number = findNodeHandle(this._animatedView);
+    invariant(nativeViewTag != null, 'Unable to locate attached view in the native tree');
+    NativeAnimatedAPI.connectAnimatedNodeToView(this.__getNativeTag(), nativeViewTag);
+    if (this._props.style) {
```
Instead of doing that, could you change the implementation of __getValue() in the case of natively implemented values to return only the static ones? I'm afraid that if it's handled out of band like this, it can introduce race conditions where the props are updated but the node is not connected anymore and it gets out of sync.

https://github.com/facebook/react-native/pull/6466/files#diff-f048d92ca0be679bc38d38147b311100R1423

vjeux commented on Libraries/Animated/src/AnimatedImplementation.js:
```diff
@@ -170,12 +277,28 @@ class TimingAnimation extends Animation {
+    }
+    return {
+      type: 'frames',
+      frames: mults,
```
if you call it frames, could you also call the variable frames?

vjeux commented on Libraries/Animated/src/AnimatedImplementation.js:
```diff
@@ -1100,17 +1277,78 @@ class AnimatedProps extends Animated {
     this._callback();
   }
+
+  __makeNative(): void {
```
I don't see you handling the following case:

```
var a1 = Animated.Value(0);
var a2 = Animated.Value(1);

style={{transform: [{translateX: a1}, {translateY: a2}]}}

Animated.timing(a1, {useNativeDriver: true}).start();
Animated.timing(a2, {useNativeDriver: false}).start();
```

A value shouldn't be able to depend on both a native and non native animation. They would otherwise be updated by both native and js, race to update and use half an incomplete value.

What we should ensure is that the entire tree is either all native or all non-native, but not a mix of both.

vjeux commented on Libraries/Animated/src/AnimatedImplementation.js:
```diff
@@ -28,20 +30,90 @@ import type { InterpolationConfigType } from 'Interpolation';
+  },
+};
+
 // Note(vjeux): this would be better as an interface but flow doesn't
```
At this point, we probably want to remove that since we have so much implementation in it :p

kmagiera replied:
I decided to add this here to have all the native module method definitions in JS (with flow types). Otherwise you just have that reference to a native module and it's difficult to tell from the JS point of view what the API looks like

vjeux commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/FrameBasedAnimationDriver.java:
```diff
@@ -0,0 +1,76 @@
+  private final double[] mFrames;
+  private final double mToValue;
+  private double mFromValue;
+  private boolean mHasToValue;
```
Why do you have this variable? You always set toValue and toValue is required.

One suggestion: it would greatly simplify the implementation if you always added the toValue as last element of the array. This an easy way to solve the problem of the last element not ending at the exact end value.

vjeux commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedModule.java:
```diff
@@ -0,0 +1,228 @@
+  }
+
+  @ReactMethod
+  public void disconnectAnimatedNodes(final int parentNodeTag, final int childNodeTag) {
```
In the JS version I'm using `attach/detach` naming convention. Is there a reason why you invent a new one here?

kmagiera replied:
not really, it's just that I cannot directly map JS attach calls to native attach calls (nodes can be build separately and only when start is called I know if the graph should be sent to native)

kmagiera replied:
I'd actually prefer to stick with the current naming convention (connect/disconnect). 
1) Those method are actually closer to `__addChild` and `__removeChild` than they are to `__attach` and `__detach`
2) As mentioned earlier they are not a immediate consequence of calling `addChild` or `removeChild` in JS which IMO it could be confusing to name them similarly.

If you think we should use existing naming convention I'd rather rename them to sth like `addChildToAnimatedNode`/`removeChildFromAnimatedNode`, what do you think?

vjeux commented:
Nice, sorry I haven't reviewed any of this earlier!

Overall, this looks really good. This is how I envisioned running it in native would look like (but never actually tried to do it :p). At this point, it's mostly going to be a matter of porting all the animation types and attributes we support to native, boring work but not difficult.

The biggest thing I would recommend changing is the way you iterate over the animated nodes. It would be nice to use the same pattern as JavaScript, which I believe is simpler and I know is working (I couldn't find a way to mentally break yours but I'm not 100% confident).

Can't wait to see it go live! I'll make sure to review all the follow ups very quickly so that we can move fast on this.

aleclarson commented:
> In JS they use `setTimeout` to run value update loop every frame.

@kmagiera It appears that `requestAnimationFrame` is what's being used (as seen [here](https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/AnimatedImplementation.js#L191)).

How far along is the iOS implementation? And who's working on it?

This PR is pure diamonds! :gem:

kmagiera commented:
@vjeux So in my opinion the top-down + bottom-up approach is not correct and topological order is exactly what we want :) (I could be wrong of course, so please see my explanation below)

The problem with calculating node updates in non-topological order can be illustrated with the following example:
```
Animation1 ---> ValueNode(0.4) ---> Interpolation ---> AnimatedStyle ---> AnimatedProperty
                                                         ^
                                                        /
Animation2 --->  ValueNode(0.2) -----------------------/
```
The approach we currently use in `AnimatedImplementation.js` is slightly different than what you describe on one of the [inline comments above](https://github.com/facebook/react-native/pull/6466#discussion_r56880276). That difference actually makes it work but adds a little overhead (we traverse the graph more times than necessary). We don't actually run all the active animations all at once but instead each of the active animation has its own timer function (created using `requestAnimationFrame` see it for example [here](https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/AnimatedImplementation.js#L191)). If we were running all the animations at once (that's how I do it in "native" actually) and doing simple top-to-bottom traversal (DFS, or however you like to call it), there is a risk we'd reach update method of `AnimatedStyle` node (which takes `Interpolation` value and `ValueNode(0.2)` as an input) before `Interpolation` is updated. That, clearly, would have an undesirable effect (we will take "stale" value of `Interpolation` node as an input and use it to update `AnimatedStyle` node and that value will propagate to `AnimatedProperty`). The reason why this works with the current implementation of Animated is because we will traverse this graph twice (once each of those two animation would fire). So the first traversal will start from `Animation1` and update `AnimatedStyle` using "stale" value from `ValueNode(0.2)`, but next one will update `ValueNode(0.2)` and will again update `AnimatedStyle` and it's children. As a result we will actually enqueue 2 calls to `setNativeProps` (which is suboptimal): one will contain "stale" value, the second one will be correct. You can see this using [this example](https://gist.github.com/kmagiera/ce986276557b3f9c26f0) and enabling `SPY_MODE` on a bridge to see that "opacity" and "translateX" value only change every two calls:

```
RKUIManager.updateView([36,"RCTView",{"opacity":0.9743458965152945,"translateX":98.54539653194766}])
RKUIManager.updateView([36,"RCTView",{"opacity":0.9845642488272184,"translateX":98.54539653194766}])
RKUIManager.updateView([36,"RCTView",{"opacity":0.9845642488272184,"translateX":98.76260607591188}])
RKUIManager.updateView([36,"RCTView",{"opacity":0.9879705818383308,"translateX":98.76260607591188}])
```

In this PR I run all the active animation updates in a single loop and thus I need to avoid calling `update` on nodes unless all their inputs has been visited (updated) --- visiting nodes in [topological order](https://en.wikipedia.org/wiki/Topological_sorting#Kahn.27s_algorithm) satisfies this criteria.

vjeux commented:
Thanks for the detailed reply! I'm now convinced that this is actually a more efficient way to execute the updates. Would be nice to update the js implementation to do it as well :)

It also made me realize that I didn't actually implement the layer that would first run all the values updates and then the flush. I made so many versions that I thought it would have been there.

kmagiera commented:
Wait @facebook-github-bot, I haven't finished with the updates

kmagiera commented:
Hey @astreet, @vjeux. Thanks for the review. I updated this PR taking into account your comments (although I might've missed one or two as there were so many of them so please don't yell at me If I forgot about something). I've also added even more class-level documentation and did some code refactoring (extracted part of the `AnimatedImplementation.js` changes into `NativeAnimatedHelper.js` and moved java code under `com.facebook.react.animated`)

bestander commented:
FYI tests are failing

astreet commented on ReactAndroid/src/main/java/com/facebook/react/animated/FrameBasedAnimationDriver.java:
```diff
@@ -0,0 +1,63 @@
+    long timeFromStartNanos = (frameTimeNanos - mStartFrameTimeNanos);
+    // frames are calculated at 60FPS, to get index by a given time offset from the start of the
+    // animation, we take the time diff in millisecond and divide it by 60 frames per 1000ms.
+    int frameIndex = (int) (timeFromStartNanos / 1000000L * 60L / 1000L);
```
You should still consider my previous comment where we just try to calculate the number of elapsed frames from the last time we ran runAnimationStep

kmagiera replied:
Not sure If I understand you correctly. I believe this approach is fine. In time-based animation (timing, decay) the frames array we get as an input should not be treated as values that should be used for each particular frame (so e.g. if we draw frame 1, then skipped frame 2 and draw at frame 3 we should also skip the content of frames array at index 2).

In other words: the value of frames[5] is the value that time-based animation will calculate if it was running in native at 5*1000/60 ms after the animation has started. It doesn't matter what the frame rate, we just use that frame table indices as they were in units of 16,6ms. Alternatively we could be sending a map that would look like:
```
frames = {
  0: 0.2,
  16.666: 0.3,
  33.333: 0.4,
  50: 0.5,
  ...
}
```

astreet replied:
So, the problem is that you have an animation sampled at 60 frames per second, but you might be rendering at 59 or 61 fps. This means that when you 'snap' to the 60fps values to show at 61/59 fps, the animation may no longer be smooth. I was suggesting just treating the frame array as a series of frames, no matter the framerate of the underlying platform -- this wouldn't completely respect the animation duration but I don't think that's horrible.

astreet replied:
You could also treat the frames as 'the value every 16.6 ms' and then interpolate the values between frames based on the exact time (instead of snapping to an exact 60fps frame value)

astreet commented on ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedModule.java:
```diff
@@ -0,0 +1,265 @@
+        }
+
+        if (mNodesManager.hasActiveAnimations()) {
+          mNodesManager.runUpdates(uiImplementation, frameTimeNanos);
```
Slightly prefer passing the UIImplementation as a constructor arg to the nodes manager instead.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,333 @@
+            child.mBFSColor = mAnimatedGraphBFSColor;
+            updatedNodesCount++;
+            nodesQueue.add(child);
+            child.update(nextNode, frameTimeNanos);
```
Why don't we just call this when we pop the child node off the queue? Seems like it would be more consistent with how we process the other nodes.

Also, why are we passing in the last parent to be visited to child.update (as nextNode)?

astreet commented on ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedNodesManager.java:
```diff
@@ -0,0 +1,333 @@
+   * sub-graph of *active* nodes. This is done by adding node to the BFS queue only if all its
+   * "predecessors" have already been visited.
+   */
+  public void runUpdates(UIImplementation uiImplementation, long frameTimeNanos) {
```
Would be sweet to have some tests for this :)

kmagiera replied:
Check out "what's coming up next" section in the PR description ("Unit tests for the graph traversal algorithm") :panda_face:

astreet commented on ReactAndroid/src/main/java/com/facebook/react/animated/PropsAnimatedNode.java:
```diff
@@ -0,0 +1,69 @@
+    }
+    uiImplementation.synchronouslyUpdateViewOnUIThread(
+      mConnectedViewTag,
+      new ReactStylesDiffMap(propsMap));
```
Since we're in a time sensitive (frame by frame) loop, it'd be nice to not have to allocate things (and thus trigger GCs) and instead reuse them. Might not matter much and could be done in a follow up

astreet commented:
Sweet, AnimatedNode looks way better now! I put inlines, otherwise looks good to ship from the Java side.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/animated/AnimatedNode.java:
```diff
@@ -0,0 +1,66 @@
+   *                          just been updated and triggered the update of current node
+   * @param frameTimeNanos frame time in nanoseconds provided to the android choreographer callback
+   */
+  public void update(AnimatedNode lastUpdatedParent, long frameTimeNanos) {
```
I mentioned this below, but I don't see why we're passing the last updated parent here: this is a pretty weird contract.

kmagiera replied:
I guess I can change that. It was there only to avoid storing references to parent nodes. I guess we are not going to be saving much on not doing that

vjeux commented:
I'm happy with this diff for a first pass, I hope that we can ship it and iterate with smaller subsequent diffs as it would make reviews easier going forward :)

astreet commented:
Ok Krzysztof, can you address the few remaining inlines and let me know when you're ready to ship?

astreet commented:
(And also make sure tests are passing)

kmagiera commented:
@astreet I updated PR addressing all the remaining inlines except from the 60FPS comment (see my response [here](https://github.com/facebook/react-native/pull/6466#discussion_r57247597))

kmagiera commented:
@bestander I've fixed one related test failure, but the remining circleci failure doesn't look related: 
```
404 no such package available : shebang-regex
```

bestander commented:
Well done! tests are green.
Sometimes npm install may fail with 404, a restart usually helps

aleclarson commented:
Is there an iOS implementation in the works?

ide commented:
> Is there an iOS implementation in the works?

Not to my knowledge, but someone at FB will probably author an Obj-C port down the road.

astreet commented:
We can address the framerate issue in a follow up, let's get this in while tests are still passing ;)

@facebook-github-bot shipit

dimitrovskif commented:
Have you guys considered adding another JS executor with it's own thread that can calculate animations, navigation transitions and other processes that require high framerate?

For example:
```javascript
const someAnimationAPI = {
   start: () => {
         AppRegistry.animationThread.send({ target: someViewID, curve:'ease', property: 'scaleX'});
   }
}
```

And the animation thread could be a JS executor that loads `animation.thread.js` as entry point:

```javascript
const transitionScheduler = new TransitionScheduler();
requestAnimationFrame(anim);

setThreadQueueReceiver((data, callback) => {
    if(data.type == 'ANIMATION'){
       // the animation will be scheduled next anim frame
       transitionScheduler.schedule({data, callback});
    }
});
```

It would be much easier for development and cross compatibility.

ide commented:
@dimitrovskif Yes it was considered but we wanted to first pursue a solution that didn't require a second JS VM heap.

dimitrovskif commented:
@ide Oh. The benefit of cross compatible native animations should be more important than the additional JS executor overhead, in my opinion. Writing code for all animation types and curves on all platforms might become messy

skevy commented:
@dimitrovskif yah the overhead of the extra JS heap is significant.

An iOS version of this is not going to be too hard to write...just needs someone to crank it out. And once it's built, it's relatively self-sustaining I think. There's still plenty of logic in JS that's shared cross-platform.

vjeux commented:
For context, right now we are having a really big effort on reducing the memory overhead of react native in order to be embedded inside of another app (the main fb app). The app by itself is already taking up a bunch of memory and the more you add ontop of that, the more likely you are going to have your app killed when you switch apps as the os wants to reclaim memory or even while the app is running.

If we want to ship there, we just cannot afford a second js heap right now. It would indeed make code sharing and development much easier though, so something we may reconsider in the future.

ms88privat commented:
I'm having problems integrating the package to my project, any help here please? 

Thats what I did try, but where do I get the `ReactApplicationContext` from?

```java
import com.facebook.react.animated.NativeAnimatedModule;

/**
     * A list of packages used by the app. If the app uses additional views
     * or modules besides the default ones, add more packages here.
     */
    @Override
    protected List<ReactPackage> getPackages() {
        mCallbackManager = new CallbackManager.Factory().create();
        return Arrays.<ReactPackage>asList(
            new MainReactPackage(),
            new VectorIconsPackage(),
            new ImagePickerPackage(),
            new FBSDKPackage(mCallbackManager),
            new ExtraDimensionsPackage(this),
            new NativeAnimatedModule(???)
        );
    }

dimitrovskif commented:
@ms88privat You are supposed to create your own package, and then add `NativeAnimatedModule` inside the package:

```java
@Override
public List<NativeModule> createNativeModules( ReactApplicationContext reactContext) {
    List<NativeModule> modules = new ArrayList<>();
    modules.add(new NativeAnimatedModule(reactContext));
    return modules;
}
```

After doing that, you may put your package inside `getPackages()`. You may read more info here: [here](https://facebook.github.io/react-native/docs/native-modules-android.html#content)

ms88privat commented:
@dimitrovskif thanks, I will try that. Another question regarding "layout animations". I got one value, which I want to animate and then interpolate this value to the properties  top, left, width, height and opacity. Would this be possible right now and result into better performance? Because at the moment the performance is completely crap on Android (even with dev false).

nschurmann commented:
@dimitrovskif hello!, could you provide an example of the package that you need to create?, my implementation fails silently:

```java
package com.polygonsmash;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;
import com.facebook.react.animated.NativeAnimatedModule;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by nicolas on 6/23/16.
 */
class NativeAnimation implements ReactPackage {

    @Override
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }

    @Override
    public List<NativeModule> createNativeModules(
            ReactApplicationContext reactContext) {
        List<NativeModule> modules = new ArrayList<>();

        modules.add(new NativeAnimatedModule(reactContext));

        return modules;
    }
}
```

```java
package com.polygonsmash;

import com.facebook.react.ReactActivity;
import com.facebook.reactnative.androidsdk.FBSDKPackage;
import com.facebook.react.ReactPackage;
import com.facebook.react.shell.MainReactPackage;

import android.content.Intent;     // <--- import
import android.os.Bundle;

import com.facebook.CallbackManager;
import com.facebook.FacebookSdk;
import com.facebook.reactnative.androidsdk.FBSDKPackage;

import com.polygonsmash.NativeAnimation;

import java.util.Arrays;
import java.util.List;

public class MainActivity extends ReactActivity {
    CallbackManager mCallbackManager;
    /**
     * Returns the name of the main component registered from JavaScript.
     * This is used to schedule rendering of the component.
     */
    @Override
    protected String getMainComponentName() {
        return "polygonsmash";
    }

    /**
     * Returns whether dev mode should be enabled.
     * This enables e.g. the dev menu.
     */
    @Override
    protected boolean getUseDeveloperSupport() {
        return BuildConfig.DEBUG;
    }

    /**
     * A list of packages used by the app. If the app uses additional views
     * or modules besides the default ones, add more packages here.
     */
    @Override
    protected List<ReactPackage> getPackages() {
        mCallbackManager = new CallbackManager.Factory().create();
        ReactPackage packages[] = new ReactPackage[]{
                new MainReactPackage(),
                new NativeAnimation(),
                new FBSDKPackage(mCallbackManager),
        };
        return Arrays.<ReactPackage>asList(packages);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mCallbackManager.onActivityResult(requestCode, resultCode, data);
    }
}
```

```javascript
animate() {
    Animated.parallel([
      Animated.spring(this.bounceValue, {
        toValue: 1,
        friction: 3,
        useNativeDriver: true, // <-- doesn't affect performance
      }),
      Animated.timing(this.pan, { 
        duration: 1500,
        easing: Easing.inOut(Easing.linear),
        toValue: -200,
        useNativeDriver: true, // <-- stops the execution of everything without throwing any error
      })
      ,
      Animated.timing(this.fadeAnim, {
        duration: 1500,
        easing: Easing.inOut(Easing.linear),
        toValue: 0,
        useNativeDriver: true, // <-- stops the execution of everything without throwing any error
      })
    ]).start()
  }
```

Introvertuous commented:
I don't mean to revive a dead thread but is this not implemented yet? Is there support for transform?

DannyvanderJagt commented:
Is there any eta/roadmap on this subject? 

React native is an awesome project and it is really useful but due to the nature of js (single threaded) the performance does not compare to truly native apps. RN could receive a huge boost when animations are off loaded to a UI thread. It would benefit the current (Experimental) Navigator (animations) since that is tremendous bottle neck at the moment. 

Would love to help.

DannyvanderJagt commented:
Never mind, after some more digging I found out that the RN team is working on this and that parts are already in RN.  See [here](https://productpains.com/post/react-native/offload-some-animations-from-js-thread-for-better-perf)
